### DISCUSSION TITLE
How to handle keys that should be hidden in some layouts?

### DISCUSSION THREAD

**Discussion by antoniusnaumann** [nickcoutsos/keymap-layout-tools#6] [General]
I am currently working on a project (little side project as of now with not much to share), which I'd like to be compatible with the keymap.json created with this tool.

Since it is possible to define multiple layouts under the layouts tag, I wonder what the correct way to implement several layout variants (e.g. 3x6 corne, 3x5 corne, etc.). From my understanding, I would create a keymap.json that looks like this, with 2 different layout variants:

```json
{
  "layouts": {
    "3x6": {
      "layout": [
        { "label":    "Tab", "row": 0, "col":  0, "x":     0, "y":   0.36 },
        { "label":      "Q", "row": 0, "col":  1, "x":     1, "y":   0.36 },
        ...
      ]
    },
    "3x5": {
      "layout": [
        { "label":      "Q", "row": 0, "col":  1, "x":     1, "y":   0.36 },
        ...
      ]
  }
}
```

But how can I reflect that when generating a ZMK keymap from the 3x5 layout, `&none` should be inserted at the positions of the omitted keys? Is this covered by the spec as of now? 

I would imagine that there should be a "hidden" attribute for those cases that indicates that a key is omitted from the layout: 
(in this case, making x and y optional is probably a good idea)
```json
...
"3x5": {
      "layout": [
        { "label":    "Tab", "row": 0, "col":  0, "hidden": true },
        { "label":      "Q", "row": 0, "col":  1, "x":     1, "y":   0.36 },
        ...
      ]
  }
...
```

**Comment by nickcoutsos** [comment]
You've got the multiple layouts idea right, but one minor nitpick for clarity: `keymap.json` is old news and the layout and other metadata are in `info.json` (or `<keyboard name>.json` to support multiple keyboards in the same repo).

> But how can I reflect that when generating a ZMK keymap from the 3x5 layout, &none should be inserted at the positions of the omitted keys? Is this covered by the spec as of now?

This is largely the point of having multiple layouts. Layout selection in my app is meant to map directly to the keyboard's matrix transform. I'll point to Corne again for an example:

In ZMK's [`corne.dtsi`](https://github.com/zmkfirmware/zmk/blob/main/app/boards/shields/corne/corne.dtsi) you'll see multiple transforms (nodes which contain a `compatible = "zmk,matrix-transform";` property) and a `chosen` node where the default is specified with `zmk,matrix_transform = &default_transform;`. That can be overridden anywhere in the devicetree so you'll commonly see in `config/corne.keymap` another chosen node where a different transform is referenced.

When I pre-generate layouts from a matrix transform I use those references as the layout name/id, but if you're manually creating them that's fine as long as the names match up. Just note that devicetree labels can be up to 31 characters from the set `a-zA-Z0-9_` but the first character must be a letter.

I'd be happy to take a look at what you have so far if you need help.

**Reply by antoniusnaumann** [reply]
Ah thanks. I overlooked that some keyboards define different matrix transforms, it totally makes sense to build on top of that.

Just to clarify: 
Is the chosen matrix transform name matched via the key in the "layouts" dictionary? Or by the optional "name" parameter? 

I'd assume the first is the case so for:

```
"layouts": {
    "corne_3x5": {
        "name": "Corne mini"
        ...
    }
}
```

I assume the matrix transform chosen will be "corne_3x5" and the "name" is just for display purposes?

Edit: I don't have much so far, I'd just like to build some utility libraries in Rust for ZMK keyboard-related things and perhaps an offline Desktop app on top of this. I figured it would be a good idea to use the same keymap layout format instead of adding one more.

**Reply by nickcoutsos** [reply]
Exactly that, although... I wonder if I even use `name` anywhere?

**Reply by antoniusnaumann** [reply]
I think I have a keymap file for a custom keyboard (that was derived automatically from the ZMK dtsi by your keymap editor) somewhere that had a "name" property set for its layout. 

At least for my use case, that is quite useful as I'd like to store a display name for different layouts somewhere.

**Reply by nickcoutsos** [reply]
Ok, fair enough. I'm starting to have my doubts about it so I don't want to make any promises but if you run into issues with a mismatched layout object key and `name` property you can make an issue and I'll hopefully fix it.
